### PR TITLE
Equation solver for Cubic & Quartic Equations

### LINKED ISSUE #190: Module for Cardano's Algorithm
Cubic equations can be implemented using [Cardano's](https://brilliant.org/wiki/cardano-method/) method.
 
_File to be modified:_ `visma/polynomial/roots.py`

### PR DESCRIPTION
**Pull request by mayankDhiman**
**Changelog:**
- Module for solving _Cubic Equations using Determinant_ based formula added
- Module for solving _Quartic Equations using Determinant_ based formula added
- `polynomial/roots.py` module restructured to incorporate equations of both types (Quadratic & Cubic)
- Test cases for Cubic equation solver added.
- A bug fixed

**Issues Addressed:**
Should fix #190 & #203 

**Examples:**
- In CLI: [Cubic](https://pastebin.com/kjEASty1) and [Quartic](https://pastebin.com/31U4npCH) 
- In GUI
Screenshot from 2019-06-27 01-12-28
Screenshot from 2019-06-27 01-11-58
Screenshot from 2019-07-03 00-26-03

### CODE REVIEW

File: tests/test_solvers.py [hunk 1]
```diff
@@ -8,31 +8,42 @@
...
-    assert quickTest("x^2 + 1 - 2x = 0", quadraticRoots) == "(x-1.0)^(2)=0"
-    assert quickTest("x^2 + 1 = 2x", quadraticRoots) == "(x-1.0)^(2)=0"
-    assert quickTest("x^2 = 2x - 1", quadraticRoots) == "(x-1.0)^(2)=0"
-    assert quickTest("-2x = - x^2 - 1", quadraticRoots) == "(x-1.0)^(2)=0"
-    # FIXME: assert quickTest("0 = 2x - x^2 - 1", quadraticRoots) == "(x-1.0)^(2)=0"
+    assert quickTest("x^2 + 1 - 2x = 0", rootFinder) == "(x-1.0)^(2)=0"
+    assert quickTest("x^2 + 1 = 2x", rootFinder) == "(x-1.0)^(2)=0"
+    assert quickTest("x^2 = 2x - 1", rootFinder) == "(x-1.0)^(2)=0"
+    assert quickTest("-2x = - x^2 - 1", rootFinder) == "(x-1.0)^(2)=0"
+    # FIXME: assert quickTest("0 = 2x - x^2 - 1", rootFinder) == "(x-1.0)^(2)=0"
```
Comment (8hantanu) [inline tests/test_solvers.py:23]: Is this still giving an error?

File: tests/test_solvers.py [hunk 2]
```diff
@@ -8,31 +8,42 @@
...
 
 
-def test_quadraticRoots():
+def test_rootFinder():
 
-    assert quickTest("x^2 + 2x + 1 = 0", quadraticRoots) == "(x+1.0)^(2)=0"
-    assert quickTest("x^2 + 2x = - 1", quadraticRoots) == "(x+1.0)^(2)=0"
-    assert quickTest("x^2 = - 2x - 1", quadraticRoots) == "(x+1.0)^(2)=0"
-    assert quickTest("0 = x^2 + 2x + 1", quadraticRoots) == "(x+1.0)^(2)=0"
+    #  Tests for Quardatic (2nd Degree) Equations
```
Comment (8hantanu) [inline tests/test_solvers.py]:
```suggestion
    #  Tests for Quadratic (2nd Degree) Equations
```

File: visma/io/checks.py [hunk 3]
```diff
@@ -267,7 +267,7 @@ def checkTypes(lTokens=None, rTokens=None):
         isPoly, polyDegree = preprocessCheckPolynomial(copy.deepcopy(lTokens), copy.deepcopy(rTokens))
         if isPoly:
             availableOperations.append('factorize')
-            if(polyDegree == 2):
+            if(polyDegree == 2 or polyDegree == 3):
```
Comment (8hantanu) [inline visma/io/checks.py]: `polyDegree == 4` case ?

File: visma/solvers/polynomial/cubic.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    coeffs = getCoefficients(lTokens, rTokens, 3)
+    var = getVariables(lTokens)
+    roots, animNew2, commentNew2 = getRootsCubic(coeffs)
+    animations.extend(animNew2)
+    comments.extend(commentNew2)
+    tokens1 = []
+    expression1 = Expression()
+    expression1.coefficient = 1
+    expression1.power = 3
+    variable = Variable()
```
Comment (8hantanu) [inline visma/solvers/polynomial/cubic.py]:
```suggestion
    variable = Variable(1, var, 1)
```
Can use the same method for initializing the `Expression` and `Constant` objects at other places. The Expression's `__init__` might need a little fixing.

**Review by 8hantanu — COMMENTED** [review]
Great work :+1: !

File: tests/test_solvers.py (hunk 1 above)
Comment (mayankDhiman) [reply tests/test_solvers.py:23]: Yes, it is (kind of stuck in a loop when this input is given). Will update if it is resolved with next updates.

File: visma/solvers/polynomial/cubic.py (hunk 4 above)
Comment (mayankDhiman) [reply visma/solvers/polynomial/cubic.py]: Done.

**Review by 8hantanu — APPROVED** [review]

**Pull request merged**